parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -337,14 +338,27 @@ pub mod pallet {
 			Ok(ticket)
 		}
 
-		fn submit(ticket: Self::Ticket) -> Result<MessageHash, SubmitError> {
-			// Make sure the bridge not halted
+		fn submit(ticket: Self::Ticket, priority: Priority) -> Result<MessageHash, SubmitError> {
 			Self::ensure_not_halted().map_err(|_| SubmitError::BridgeHalted)?;
-			T::MessageQueue::enqueue_message(
-				ticket.message.as_bounded_slice(),
-				AggregateMessageOrigin::Parachain(ticket.origin),
-			);
-			Self::deposit_event(Event::MessageQueued { id: ticket.id });
+			match priority {
+				Priority::Normal => {
+					T::MessageQueue::enqueue_message(
+						ticket.message.as_bounded_slice(),
+						AggregateMessageOrigin::Parachain(ticket.origin),
+					);
+					Self::deposit_event(Event::MessageQueued { id: ticket.id });
+				},
+				Priority::High => {
+					ensure!(
+						MessageLeaves::<T>::decode_len().unwrap_or(0) <
```

Comment:
So unfortunately, if there is congestion, then this `ensure!` check is always going to fail. The MessageQueue pallet is going to fill up `MessageLeaves` at the start of every block, and there won't be any more space for `High` priority messages.

Reply:
Not sure but I would assume the message is queued by [Origin](https://github.com/paritytech/polkadot-sdk/blob/2c35bc3d273f47d45c17078134c048016ef7a57a/substrate/frame/message-queue/src/lib.rs#L766) and the congestion for normal user operations(e.g. sendToken from assetHub) will not block the control commands here origin from bridgeHub.

And I would prefer these low level of the QoS priority control of traffic stuff to be addressed by the internal implementation of `message-queue` pallet.

Reply:
Since it's not easy for the upstream `message-queue` to support priority control natively. After some more thought revamp in https://github.com/Snowfork/snowbridge/pull/949/commits/fe5c8a3d96a7a5ac4f68f0339d0b655eea8287fe
from our side for the feature `swap low priority message with high priority one when hard limit reached`.

Reply:
Unfortunately, that is not going to work, and will result in high priority messages being dropped. See my other comment about it.

Reply:
Thanks for pointing it out. Yes, I also realize it and revamp in https://github.com/Snowfork/snowbridge/pull/949/commits/42692ab2fd457618d325fb7ce87e89f08bd42c8f